enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,511 @@
+		return eg.Wait()
+	})
+	if err != nil {
+		logger.Error(ctx, "failed to reconcile", slog.Error(err))
+	}
+
+	return err
+}
+
+// SnapshotState captures the current state of all prebuilds across templates.
+func (c *StoreReconciler) SnapshotState(ctx context.Context, store database.Store) (*prebuilds.GlobalSnapshot, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	var state prebuilds.GlobalSnapshot
+
+	err := store.InTx(func(db database.Store) error {
+		presetsWithPrebuilds, err := db.GetTemplatePresetsWithPrebuilds(ctx, uuid.NullUUID{}) // TODO: implement template-specific reconciliations later
+		if err != nil {
+			return xerrors.Errorf("failed to get template presets with prebuilds: %w", err)
+		}
+		if len(presetsWithPrebuilds) == 0 {
+			return nil
+		}
+		allRunningPrebuilds, err := db.GetRunningPrebuiltWorkspaces(ctx)
+		if err != nil {
+			return xerrors.Errorf("failed to get running prebuilds: %w", err)
+		}
+
+		allPrebuildsInProgress, err := db.CountInProgressPrebuilds(ctx)
+		if err != nil {
+			return xerrors.Errorf("failed to get prebuilds in progress: %w", err)
+		}
+
+		presetsBackoff, err := db.GetPresetsBackoff(ctx, c.clock.Now().Add(-c.cfg.ReconciliationBackoffLookback.Value()))
+		if err != nil {
+			return xerrors.Errorf("failed to get backoffs for presets: %w", err)
+		}
+
+		state = prebuilds.NewGlobalSnapshot(presetsWithPrebuilds, allRunningPrebuilds, allPrebuildsInProgress, presetsBackoff)
+		return nil
+	}, &database.TxOptions{
+		Isolation:    sql.LevelRepeatableRead, // This mirrors the MVCC snapshotting Postgres does when using CTEs
+		ReadOnly:     true,
+		TxIdentifier: "prebuilds_state_determination",
+	})
+
+	return &state, err
+}
+
+func (c *StoreReconciler) ReconcilePreset(ctx context.Context, ps prebuilds.PresetSnapshot) error {
+	logger := c.logger.With(
+		slog.F("template_id", ps.Preset.TemplateID.String()),
+		slog.F("template_name", ps.Preset.TemplateName),
+		slog.F("template_version_id", ps.Preset.TemplateVersionID),
+		slog.F("template_version_name", ps.Preset.TemplateVersionName),
+		slog.F("preset_id", ps.Preset.ID),
+		slog.F("preset_name", ps.Preset.Name),
+	)
+
+	actions, err := c.CalculateActions(ctx, ps)
+	if err != nil {
+		logger.Error(ctx, "failed to calculate actions for preset", slog.Error(err), slog.F("preset_id", ps.Preset.ID))
+		return nil
+	}
+
+	// nolint:gocritic // ReconcilePreset needs Prebuilds Orchestrator permissions.
+	prebuildsCtx := dbauthz.AsPrebuildsOrchestrator(ctx)
+
+	levelFn := logger.Debug
+	switch actions.ActionType {
+	case prebuilds.ActionTypeBackoff:
+		levelFn = logger.Warn
+	case prebuilds.ActionTypeCreate, prebuilds.ActionTypeDelete:
+		// Log at info level when there's a change to be effected.
+		levelFn = logger.Info
+	}
+
+	fields := []any{
+		slog.F("action_type", actions.ActionType),
+		slog.F("create_count", actions.Create),
+		slog.F("delete_count", len(actions.DeleteIDs)),
```

Comment:
I think it's a pity that we don't log the current state of the preset here anymore. This will be helpful for us when debugging reconciliation problems. What is the reason why you removed these fields?

https://github.com/coder/coder/blob/dk/prebuilds/enterprise/coderd/prebuilds/reconcile.go#L307-L314

Reply:
Okay, I'll put it back. Initially I removed it because I splitted Actions & State structures. Basically I only need 3 fields to make a reconciliation decision:
- backoff
- create
- deleteIDs

I don't need like 12 fields including `outdated`, `extraneous`, `failed`, etc... Because it maybe a bit confusing for API consumer if I see structure like this:
https://github.com/coder/coder/commit/fe60b569ad754245e28bac71e0ef3c83536631bb#diff-e435a5b0fb81846382c9401a9058a87a473f9f39eca0e884d1dd981c4d2885a1R29-R41

Reply:
@dannykopping I put it back.